File: Astro-TFG/public/scripts/4-manejo-peticiones/prueba-2/version-1/javascript/main.js
let worker = null;
let workerTime = 0;

function initializeWorker() {
    if (!worker) {
        worker = new Worker(
            window.location.origin +
            window.document.body.dataset.jsPath +
            "worker.js"
        );
    }
}

async function _runJsBenchmark() {
    try {
        window.clearCell("javascript-output");
        window.showExecutionLoader();

        const workerStartTime = performance.now();
        await initializeWorker();
        const workerTime = performance.now() - workerStartTime;

        const numRequests = parseInt(
            document.getElementById("num-requests-javascript").value,
            10
        );

        const delay = parseInt(
            document.getElementById("request-delay-javascript").value,
            10
        );

        const id = `js-${Date.now()}`;

        let apiUrl = location.origin + "/api/4.2.1/socket";

        const resultJson = await new Promise((resolve, reject) => {
            function onMessage(e) {
                if (e.data.id !== id) return;
                worker.removeEventListener("message", onMessage);
                if (e.data.error) {
                    reject(new Error(e.data.error));
                } else if (e.data.json !== undefined) {
                    resolve(e.data.json);
                } else {
                    reject(new Error("Worker dont return results"));
                }
            }
            worker.addEventListener("message", onMessage);
            worker.postMessage({
                id: id,
                num_requests: numRequests,
                delay: delay,
                api_url: apiUrl,
            });
        });

        const result = JSON.parse(resultJson);
        displayResult(result);

    } catch (error) {
        console.error(error);
    } finally {
        window.hideExecutionLoader();
    }
}

function createDiv() {
    const div = document.createElement("div");
    return div;
}

function displayResult(r) {
    const output = document.getElementById("javascript-output");

    const workerDiv = createDiv();
    workerDiv.textContent = `Worker init time: ${workerTime.toFixed(2)} ms`;
    output.appendChild(workerDiv);

    const avgTimeDiv = createDiv();
    avgTimeDiv.textContent = `Avg request time: ${r.average_time_ms.toFixed(2)} ms`;
    output.appendChild(avgTimeDiv);

    const totalTimeDiv = createDiv();
    totalTimeDiv.textContent = `Total ET: ${r.total_time_ms.toFixed(2)} ms`;
    output.appendChild(totalTimeDiv);

    const requestsDiv = createDiv();
    requestsDiv.textContent = `Requests: ${r.total_requests}`;
    output.appendChild(requestsDiv);

    const lastValueDiv = createDiv();
    lastValueDiv.textContent = `Last value: ${r.last_value.toFixed(2)}`;
    output.appendChild(lastValueDiv);
}

window.runJsBenchmark = _runJsBenchmark;

File: Astro-TFG/public/scripts/4-manejo-peticiones/prueba-2/version-1/python/main.py
# main.py
from pyscript import display, document, PyWorker
import js  # type: ignore
import time
import asyncio

worker = None


def update_results(res, worker_time):
    display(f"Worker Init Time: {worker_time:.2f} ms",
            target="pyscript-output")
    display(
        f"Avg Request Time: {res.avg_time:.2f} ms", target="pyscript-output")
    display(f"Total Time: {res.total_time:.2f} ms",
            target="pyscript-output")
    display(
        f"Finished Requests: {len(res.individual_times)}", target="pyscript-output")
    last = res.results[-1]
    display(
        f"Last Result Value: {last.data[0].value}", target="pyscript-output")


async def launch_worker(event):
    global worker
    t0 = time.perf_counter()
    if worker is None:
        base = js.window.location.origin + js.window.document.body.dataset.pyPath
        worker = PyWorker(
            f"{base}worker.py",
            type="pyodide",
            config=f"{base.replace('python', 'json')}pyscript-worker.json"
        )
        await worker.ready
    worker_time = (time.perf_counter() - t0) * 1000  # en ms

    try:
        num_requests = int(document.getElementById(
            "num-requests-pyscript").value)
        delay = int(document.getElementById("request-delay-pyscript").value)

        res = await worker.sync.do_ws_requests(num_requests, delay)
        print(res.avg_time)
        update_results(res, worker_time)
        js.window.hideExecutionLoader()
    except Exception as e:
        display(f"Error: {e}", target="pyscript-output")


def js_run_py_benchmark(event):
    js.clearCell('pyscript-output')
    asyncio.ensure_future(launch_worker(None))


js.window.run_py_benchmark = js_run_py_benchmark
